module/zfs/zfs_log.c

```diff
@@ -499,6 +500,94 @@ zfs_log_rename(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype,
 	zil_itx_assign(zilog, itx, tx);
 }
 
+/*
+ * At the moment, only Linux supports renameat2 variant of renameat, which
+ * adds three new flags of interest for us:
+ *     RENAME_NOREPLACE: if the target name at the moment of the call exists,
+ *                       don't rewrite it and return error
+ *     RENAME_EXCHANGE: atomically swap the two names on the filesystem
+ *     RENAME_WHITEOUT: creates a whiteout inode in place of renamed file as
+ *                      an atomic operation
+ *
+ * Ideally, these operations should be represented as new ZFS Intent Log record
+ * types, which should mandate a new ZFS feature flag due to the on-disk format
+ * change. One would use spa_feature_incr/decr functions to indicate that we're
+ * actually actively using the new on-disk txtypes - but these functions are
+ * only supposed to be called from the txg syncing context.
+ * This means that would need to force out in-progress txg to disk and start
+ * a new one before writing any ZIL records, just so we can be sure that ZIL
+ * replaying ZFS gets told it should expect potentially incompatible ZIL
+ * txtypes. Doing this would hurt performance.
+ * Alternatively, we could just activate the feature on a pool when these
+ * renameat2 flags get first used and leave it at that - which would render the
+ * pool importable read-only on implementations without the new feature flag,
+ * even when no new txtypes IL records would be present on-disk - which on most
+ * setups could be 'almost' all the time, so it'd be a shame to have them all
+ * read-only on non-Linux platforms.
+ * As a third option, at least until more platforms implement renameat2, we
+ * choose to rely on the fact that the ZIL is replayed in single-threaded mode
+ * before the dataset is mounted. This way, we can represent the otherwise
+ * atomic operations as a series of plain good old txtypes known to all current
+ * OpenZFS implementations. To do that, we use these hacky functions:
+ *
+ * zfs_log_rename_exchange
+ * zfs_log_rename_whiteout
+ *
+ *     To represent atomic rename with old non-atomic operations, we need
+ *     a temporary new name; so we try picking a name until we succeed, then
+ *     we get a dirent lock for that temp name until the final itx gets queued
+ */
+
+void
+zfs_log_rename_exchange(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype,
+    znode_t *sdzp, char *sname, znode_t *tdzp, char *dname, znode_t *szp)
+{
+	zfs_dirlock_t *tmpdl;
+	znode_t *tmpzp = NULL;
+	char rndname[16];
+	char *tmpname;
+	int retries = 0;
+	int error;
+
+	tmpname = kmem_alloc(MAXPATHLEN, KM_SLEEP);
+	ASSERT3P(tmpname, !=, NULL);
+
+	for (int i = 0; i < 16; i++) {
+		int r = 0xFF;
+		while (r > 127 || r == 0 || r == '/')
+			random_get_pseudo_bytes((void *)&r, 1);
+		rndname[i] = (char)r;
+	}
+
+	do {
+		retries++;
+		(void) snprintf(tmpname, MAXPATHLEN,
+		    "%s.zfs_renameat2_emul_%s%4d",
```

Comment:
I'd argue this should be a hidden file.

Reply:
It will never be presented to the user anyway, going with the spirit of the rest of the review comments, I guess I would prefer not to add logic which is not needed, what do you think?

Reply:
You don't need to add logic -- my point was that it should be prefixed with a `.` but it's only a nit.